crates/swc_ecma_transforms_compat/src/es2017/async_to_generator.rs

```diff
@@ -44,61 +44,56 @@ struct Actual {
     extra_stmts: Vec<Stmt>,
 }
 
-/// TODO: VisitMut
 #[fast_path(ShouldWork)]
-impl Fold for AsyncToGenerator {
-    noop_fold_type!();
+impl VisitMut for AsyncToGenerator {
+    noop_visit_mut_type!();
 
-    fn fold_module_items(&mut self, n: Vec<ModuleItem>) -> Vec<ModuleItem> {
-        self.fold_stmt_like(n)
+    fn visit_mut_module_items(&mut self, n: &mut Vec<ModuleItem>) {
+        self.visit_mut_stmt_like(n);
     }
 
-    fn fold_stmts(&mut self, n: Vec<Stmt>) -> Vec<Stmt> {
-        self.fold_stmt_like(n)
+    fn visit_mut_stmts(&mut self, n: &mut Vec<Stmt>) {
+        self.visit_mut_stmt_like(n);
     }
 }
 
 impl AsyncToGenerator {
-    fn fold_stmt_like<T>(&mut self, stmts: Vec<T>) -> Vec<T>
+    fn visit_mut_stmt_like<T>(&mut self, stmts: &mut Vec<T>)
     where
-        T: StmtLike + FoldWith<Actual>,
-        Vec<T>: FoldWith<Self>,
+        T: StmtLike + VisitMutWith<Actual>,
+        Vec<T>: VisitMutWith<Self>,
     {
-        let stmts = stmts.fold_children_with(self);
+        stmts.visit_mut_children_with(self);
 
-        let mut buf = Vec::with_capacity(stmts.len());
+        let stmts_len = stmts.len();
 
-        for stmt in stmts {
+        for idx in 0..stmts_len {
             let mut actual = Actual {
                 in_object_prop: false,
                 in_prototype_assignment: false,
                 extra_stmts: vec![],
             };
-            let stmt = stmt.fold_with(&mut actual);
 
-            buf.extend(actual.extra_stmts.into_iter().map(T::from_stmt));
-            buf.push(stmt);
+            stmts[idx].visit_mut_with(&mut actual);
```

Comment:
I think this is still problematic, as `idx` is not modified when `stmts` is extended.

Reply:
If `stmts.len() == 3` (stmt 0, 1, 2) and `stmt 0` creates `extra_stmts`, this code will visit `stmts[1]` which is content of `extra_stmts` from `stmt 0`.

Reply:
Yes, that's correct oversight I missed. I wanted to reuse existing vec without having separate one, but it seems bit tricky as extra_stmts are non deterministic variable.

Reply:
You can add the length of `extra_stmts` to `idx`.

Reply:
:bulb:

Reply:
actually, when loop starts ending index is already binded.. :thinking: 

I'll try to accept cost by having another vec: I think those micro optimization can come later once logic works correctly.

Reply:
Oh, yes you are right. You can add same value (the length) to end index, too.
But recreating vector is fine for now anyway.